{% extends 'layout/layout' %}

{% block content %}
    <h1>A 404 error occurred</h1>
    <h2>{{ message }}</h2>

    {% if reason is defined and reason %}

        {% if reason == 'error-controller-cannot-dispatch' %}
            {% set reasonMessage = 'The requested controller was unable to dispatch the request.' %}
        {% elseif reason ==  'error-controller-not-found' %}
            {% set reasonMessage = 'The requested controller could not be mapped to an existing controller class.' %}
        {% elseif reason ==  'error-controller-invalid' %}
            {% set reasonMessage = 'The requested controller was not dispatchable.' %}
        {% elseif reason ==  'error-router-no-match' %}
            {% set reasonMessage = 'The requested URL could not be matched by routing.' %}
        {% else %}
            {% set reasonMessage = 'We cannot determine at this time why a 404 was generated.' %}
        {% endif %}
        <p>{{ reasonMessage }}</p>

    {% endif %}

    {% if controller is defined and controller %}
        <dl>
            <dt>Controller:</dt>
            <dd>{{ escapeHtml(controller) }}
                {% if controller_class is defined and controller_class and controller_class!=controller %}
                    {{ '(' ~ 'resolves to ' ~ escapeHtml(controller_class) ~ ')' }}
                {% endif %}
            </dd>
        </dl>

    {% endif %}

    {% if display_exceptions is defined and display_exceptions %}

        {% if exception is defined and (instanceOf(exception,'Exception') or instanceOf(exception,'Error')) %}
            <hr/>
            <h2>Additional information:</h2>
            <h3>{{ getClass(exception) }}</h3>
            <dl>
                <dt>File:</dt>
                <dd>
                    <pre class="prettyprint linenums">{{ exception.getFile() }}:{{ exception.getLine() }}</pre>
                </dd>
                <dt>Message:</dt>
                <dd>
                    <pre class="prettyprint linenums">{{ escapeHtml(exception.getMessage()) }}</pre>
                </dd>
                <dt>Stack trace:</dt>
                <dd>
                    <pre class="prettyprint linenums">{{ escapeHtml(exception.getTraceAsString()) }}</pre>
                </dd>
            </dl>
            {% set previous = exception.getPrevious() %}
            {% if previous %}
                <hr/>
                <h2>Previous exceptions:</h2>
                <ul class="unstyled">
                    {% for i in 1..49 if previous %}
                        <li>
                            <h3>{{ getClass(previous) }}</h3>
                            <dl>
                                <dt>File:</dt>
                                <dd>
                                <pre class="prettyprint linenums">{{ previous.getFile() }}
                                    :{{ previous.getLine() }}</pre>
                                </dd>
                                <dt>Message:</dt>
                                <dd>
                                    <pre class="prettyprint linenums">{{ escapeHtml(previous.getMessage()) }}</pre>
                                </dd>
                                <dt>Stack trace:</dt>
                                <dd>
                                    <pre class="prettyprint linenums">{{ escapeHtml(previous.getTraceAsString()) }}</pre>
                                </dd>
                            </dl>
                        </li>
                        {% set previous = previous.getPrevious() %}
                    {% else %}
                        <li>There may be more exceptions, but we have no enough memory to proccess it.</li>
                    {% endfor %}
                </ul>
            {% endif %}

        {% else %}

            <h3>No Exception available</h3>

        {% endif %}

    {% endif %}
{% endblock %}